### PR TITLE
Reinterpret (with unsigned) instead of converting to allow for negative type parameters

### PR DESCRIPTION
**Pull request by andreasnoack**
I'm not really sure if this is the right fix but it did the job for the example I worked on, i.e. wrapping matrices from the Eigen library. For arbitrary size matrices they use `-1` as type parameter value so I got a `InexactError` when `specialize_template` tried to `convert` to a `UInt64`.

### CODE REVIEW

File: src/bootstrap.cpp [hunk 1]
```diff
@@ -372,7 +372,7 @@ JL_DLLEXPORT void *lookup_name(C, char *name, clang::DeclContext *ctx)
     return R.empty() ? NULL : R.getRepresentativeDecl();
 }
 
-JL_DLLEXPORT void *SpecializeClass(C, clang::ClassTemplateDecl *tmplt, void **types, uint64_t *integralValues,int8_t *integralValuePresent, size_t nargs)
+JL_DLLEXPORT void *SpecializeClass(C, clang::ClassTemplateDecl *tmplt, void **types, int64_t *integralValues,int8_t *integralValuePresent, size_t nargs)
```
Comment (Keno) [inline src/bootstrap.cpp]: Probably best to keep this as uint64_t and reinterpret to it on the julia side.

**Comment by Keno** [comment]
Add a test with both a negative and a large positive value?

**Comment by andreasnoack** [comment]
How about this solution? It doesn't affect very large integer values and we have this kind of *abstract reinterpretation* defined in base for the floating point type anyway, e.g.
```jl
julia> reinterpret(Unsigned, 1.0)
0x3ff0000000000000
```
If you are okay with something like this I can try to write a few tests.

**Comment by Keno** [comment]
I'm ok with this, but extending reinterpret here doesn't seem right. We should either put that into base or use a different method that falls back to `reinterpret`.

**Comment by andreasnoack** [comment]
I'll make it an unexported `_reinterpret` for now and open a pull request against base to see if it will fly there.

**Comment by Keno** [comment]
Sounds good!

**Comment by andreasnoack** [comment]
I realized that we have `unsigned` which is doing exactly the same as my `_reinterpred(Unsigned,x)`.

It was a good idea with tests because anything outside `0:255` gave wrong results because `APSInt` was set to 8 bits and not bytes. That truncation took me a while to find.

**Comment by andreasnoack** [comment]
Bump.

**Comment by Keno** [comment]
LGTM, thanks.
